### PR TITLE
Start external services before we start plugins

### PR DESCRIPTION
**Pull request by antonydenyer**
## PR description - start external services before we start plugins

Reasons:

- you may wish to query the current state of your blockchain using rpc before you start processing new blocks

- you may wish to register a websocket e.g to listen for blocks currently, you will probably miss one because we start processing before we start the ws service

To enable this a 'beforeExternalServices' hook was addedd to allow plugins to do things before those services are started. 

Signed-off-by: Antony Denyer <git@antonydenyer.co.uk>

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by macfarla — COMMENTED** [review]
LGTM

File: besu/src/main/java/org/hyperledger/besu/services/BesuPluginContextImpl.java [hunk 1]
```diff
@@ -126,13 +128,43 @@ private void addPluginVersion(final BesuPlugin plugin) {
...
+
+      try {
+        plugin.beforeExternalServices();
+        LOG.debug(
+            "beforeExternalServices called on plugin of type {}.", plugin.getClass().getName());
+      } catch (final Exception e) {
+        LOG.error(
+            "Error calling `beforeExternalServices` on plugin of type "
+                + plugin.getClass().getName()
+                + ", stop will not be called.",
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/services/BesuPluginContextImpl.java:151]: what does it mean "stop will not be called"

File: besu/src/main/java/org/hyperledger/besu/Runner.java [hunk 2]
```diff
@@ -103,31 +103,35 @@
             cacher -> new AutoTransactionLogBloomCachingService(blockchain, cacher));
   }
 
-  public void start() {
+  public void startExternalServices() {
+    LOG.info("Starting Ethereum services ... ");
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/Runner.java:107]: do we want to say "Ethereum external services" here

File: besu/src/main/java/org/hyperledger/besu/services/BesuPluginContextImpl.java (hunk 1 above)
Comment (antonydenyer) [reply besu/src/main/java/org/hyperledger/besu/services/BesuPluginContextImpl.java:151]: The plugin doesn't get added to the list of running plugins, so we don't call stop when we shutdown besu. What we're saying is you need to take care of cleaning up your own resources.

**Review by macfarla — APPROVED** [review]

**Pull request merged**